### PR TITLE
refactor: split REST module and fix github action

### PR DESCRIPTION
**Pull request by edalex-yinzi**
##### Checklist

- [x] the [contributor license agreement][] is signed
- [x] commit message follows [commit guidelines][]
- [ ] tests are included
- [ ] screenshots are included showing significant UI changes
- [x] documentation is changed or added

##### Description of change

Based on [this PR](https://github.com/openequella/openEQUELLA/pull/4676) I prefer to split the rest module because this is why we created the REST module in the first place.

1. Add missing deps for rest module
2. Separated the build rest module step from the build react front-end script  (npm prepare script).
3. Update ci
4. Update README
5. Update sbt script

[contributor license agreement]: https://www.apereo.org/node/676
[commit guidelines]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

**Comment by PenghaiZhang** [comment]
When you said `split REST module`, I was thinking you were going to create a new repo for the module.

**Comment by edalex-yinzi** [comment]
> When you said `split REST module`, I was thinking you were going to create a new repo for the module.

Ah, that might be the final step for `split` process, but for now, I think we are aiming to split the module logically first rather than physically.

**Comment by PenghaiZhang** [comment]
> > When you said `split REST module`, I was thinking you were going to create a new repo for the module.
> 
> Ah, that might be the final step for `split` process, but for now, I think we are aiming to split the module logically first rather than physically.

But with your changes, `oeq-ts-rest-api` is coupled deeper into sbt world...

**Comment by edalex-yinzi** [comment]
> > > When you said `split REST module`, I was thinking you were going to create a new repo for the module.
> > 
> > 
> > Ah, that might be the final step for `split` process, but for now, I think we are aiming to split the module logically first rather than physically.
> 
> But with your changes, `oeq-ts-rest-api` is coupled deeper into sbt world...

I think the idea is more about splitting from react front end ?

**Comment by edalex-yinzi** [comment]
@PenghaiZhang can you rerun the task? Thank you.

**Comment by edalex-yinzi** [comment]
14 is passed ！！ I will try 13.

**Comment by edalex-ian** [comment]
Okay, so it works on 14, 13 and 12. Could you please try 11 on more time @edalex-yinzi ?

**Comment by edalex-yinzi** [comment]
11 failed. @edalex-ian should I keep 12 or revert back to 10?

**Comment by edalex-ian** [comment]
> 11 failed. @edalex-ian should I keep 12 or revert back to 10?

Wow, well that definitely confirms it.

Please:

1. Go back to 12, and then we can merge once that's green again; and
2. Once that is merged, please create another PR which has it set to 11 and we'll continue from there (to look at what's wrong with PSQL on JDK 11).

Call out if that's unclear. :wink:

**Comment by edalex-yinzi** [comment]
@edalex-ian Thanks. It's ready for review.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -286,15 +286,22 @@ The New UI consists of two modules:
...
-directory and execute:
+For the most part, if you wish to work on the new UI, you need go into `oeq-ts-rest-api` directory and execute:
+
+```bash
+npm ci
+npm run build
+````
+
+after building the REST module,
+then you can go into the `react-front-end` directory and execute:
```
Comment (edalex-ian) [inline CONTRIBUTING.md:297]:
```suggestion
After building the REST module (`oeq-ts-rest-api`), then you can go into the `react-front-end` directory and execute:
```

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -286,15 +286,22 @@ The New UI consists of two modules:
...
 
 ```bash
 npm ci
 npm run dev
 ```
 
-The first will also trigger a build of the REST module, and then `run dev` will build the React App
+the `run dev` will build the React App
 and place the output into the directory from which openEQUELLA serves its web resources.
```
Comment (edalex-ian) [inline CONTRIBUTING.md:305]:
```suggestion
The `run dev` command will build the React App and place the output into the directory from which
openEQUELLA serves its web resources.
```

File: Source/Plugins/Core/com.equella.core/build.sbt [hunk 3]
```diff
@@ -135,4 +135,5 @@ clean := {
   (inplaceEditorJar / clean).value
   val baseSwagger = baseDirectory.value / "swaggerui"
   Common.nodeScript("clean", baseSwagger)
+  Common.nodeScript("clean", oeqTsRestApiDir.value)
```
Comment (edalex-ian) [inline Source/Plugins/Core/com.equella.core/build.sbt:138]: This seems miss-placed. I'd expect it to be alongside the clean for the `reactFrontEndDir`.

File: oeq-ts-rest-api/package.json [hunk 4]
```diff
@@ -32,28 +31,33 @@
...
     "axios-cookiejar-support": "^4.0.0",
-    "lodash": "^4.17.15",
-    "query-string": "^7.0.0",
-    "runtypes": "6.6.0",
-    "tough-cookie": "^4.0.0",
     "fp-ts": "2.13.1",
     "io-ts": "2.2.20",
-    "newtype-ts": "0.3.5",
+    "io-ts-types": "0.5.19",
+    "lodash": "4.17.15",
```
Comment (edalex-ian) [inline oeq-ts-rest-api/package.json:37]: Why was this version (and a few others) pinned? (i.e. why was `^` version specifier removed?)

File: project/CommonSettings.scala [hunk 5]
```diff
@@ -29,6 +29,7 @@ object CommonSettings extends AutoPlugin {
     lazy val writeScriptingJavadoc = taskKey[File]("Write the scripting javadoc")
     lazy val mergeJPF              = inputKey[Unit]("Merge all")
     lazy val buildReactFrontEnd    = taskKey[File]("Build the ReactJS based front-end")
+    lazy val oeqTsRestApiDir       = settingKey[File]("The base directory of the oEQ REST API project")
```
Comment (edalex-ian) [inline project/CommonSettings.scala:32]:
```suggestion
    lazy val oeqTsRestApiDir       = settingKey[File]("The base directory of the oEQ Typescript REST module project")
```

**Review by edalex-ian — COMMENTED** [review]
Thanks @edalex-yinzi . A few tweaks needed, plus a query about the changes to dependencies in the REST module.

File: oeq-ts-rest-api/package.json (hunk 4 above)
Comment (edalex-yinzi) [reply oeq-ts-rest-api/package.json:37]: I think we always use pinned version, so I remove the prefix '^'.

**Comment by edalex-yinzi** [comment]
Thanks @edalex-ian, updated.

File: oeq-ts-rest-api/package.json (hunk 4 above)
Comment (edalex-ian) [reply oeq-ts-rest-api/package.json:37]: Aha! That is correct, however not for this module which is designed to be consumed in another module or App and thereby we prefer to have flexibility with what may be provided by the App. :wink: 

So if you had no need to change them, then best leave as is.

**Review by edalex-ian — COMMENTED** [review]
Cool, just need you to fix up the changes for dependencies.

File: oeq-ts-rest-api/package.json (hunk 4 above)
Comment (edalex-yinzi) [reply oeq-ts-rest-api/package.json:37]: Gocha, What about the new deps ?  @edalex-ian

Comment (edalex-ian) [reply oeq-ts-rest-api/package.json:37]: It's a bit hard to determine which new ones were added. But generally they should be ranged in `dependencies` and pinned in `devDependencies`.

**Comment by edalex-yinzi** [comment]
Thanks for your help @edalex-ian , I reverted the `^` changes for old deps. And for oeq-ts-types I think it's better to keep consistency with the deps `io-ts` so I use pinned version.

**Review by edalex-ian — APPROVED** [review]
Excellent, thanks @edalex-yinzi . let's get a green build and merge, and then we'll get you to create a PR which includes PSQL 11 back in the matrix - and then we can look at figuring out what's wrong there.

**Comment by edalex-yinzi** [comment]
Flaky test failed.

**Pull request merged**